fab/fab/fabfile.py

```diff
@@ -167,107 +160,59 @@ def get_env_dict(path):
+    vars.update(get_env_dict(os.path.join(REPO_BASE, 'environmental-defaults', 'fab-settings.yml')))
+    vars.update(get_env_dict(os.path.join(REPO_BASE, 'environments', env_name, 'app-processes.yml')))
+    vars.update(get_env_dict(os.path.join(REPO_BASE, 'environments', env_name, 'fab-settings.yml')))
+    # Variables that were already in `env`
+    # take precedence over variables set in app-processes.yml
+    # except a short blacklist that we expect app-processes.yml vars to overwrite
+    overlap = set(vars_not_to_overwrite) & set(vars)
+    for key in overlap:
+        print('NOTE: ignoring app-processes.yml var {}={!r}. Using value {!r} instead.'.format(key, vars[key], vars_not_to_overwrite[key]))
+    vars.update(vars_not_to_overwrite)
+    env.update(vars)
+
+
+def _setup_env(env_name):
     env.env_name = env_name
-    env.force = force  # don't worry about kafka checkpoints if True
     env.inventory = os.path.join(REPO_BASE, 'environments', env_name, 'inventory.ini')
     load_env(env_name)
+    _confirm_branch(env.default_branch)
+    _confirm_environment_time(env_name)
     execute(env_common)
 
 
-def _confirm_branch(default_branch=None):
-    if env.code_branch == 'master':
-        if default_branch:
-            env.code_branch = default_branch
-            print ("using default branch of {}. you can override this "
-                   "with --set code_branch=<branch>".format(default_branch))
-    else:
+def _confirm_branch(default_branch='master'):
+    if not hasattr(env, 'code_branch'):
+        print("code_branch not specified, using '{}'. "
+              "You can override this with '--set code_branch=<branch>'"
+              .format(default_branch))
+        env.code_branch = default_branch
+
+    if env.code_branch != default_branch:
         branch_message = (
-            "Woah there bud! You're using branch {env.code_branch}. "
+            "Whoa there bud! You're using branch {env.code_branch}. "
             "ARE YOU DOING SOMETHING EXCEPTIONAL THAT WARRANTS THIS?"
         ).format(env=env)
         if not console.confirm(branch_message, default=False):
             utils.abort('Action aborted.')
 
 
-def _confirm_environment_time(env_name, env_tz):
-    env_hours_start_end = {
-        'enikshay': (2, 7),  # call center is offline 2am-8am IST
-        'icds': (0, 7),
-        'icds-new': (0, 7),
-    }
-    hour_start, hour_end = env_hours_start_end[env_name]
-    d = datetime.datetime.now(pytz.timezone(env_tz))
-    if hour_start <= d.hour < hour_end:
+def _confirm_environment_time(env_name):
+    window = env.acceptable_maintenance_window
+    if window:
+        d = datetime.datetime.now(pytz.timezone(window['timezone']))
+        if window['hour_start'] <= d.hour < window['hour_end']:
+            return
+    else:
         return
 
     message = (
-        "Woah there bud! You're deploying '%s' during the day. "
+        "Whoa there bud! You're deploying '%s' during the day. "
```

Comment:
[the more you know](https://www.grammarly.com/blog/whoa-woah/)

Reply:
It apparently even has a [known etymology](https://www.etymonline.com/word/whoa). You learn interesting things as a programmer.

Reply:
whoa

Reply:
![](https://store.bbcomcdn.com/images/store/skuimage/sku_RSP3720006/image_skuRSP3720006_largeImage_X_450_white.jpg)